[Sig Events] Add architecture and operator docs (#7497)
elastic/docs-content: solutions/observability/streams/significant-events/how-it-works.md

---
navigation_title: How it works
description: Pipeline internals for engineers and adopters — KI extraction, rule generation, detection, discovery, storage model, and ES|QL traceability.
applies_to:

>   serverless: beta

Reviewer: I could be wrong, but I'm fairly certain that we shouldn't use `beta` anymore when describing a lifecycle state. Here are the resources I'm referencing for guidance:
- [Versioning and availability](https://www.elastic.co/docs/get-started/versioning-availability#lifecycle-states)
- [Feature Lifecycle and Support Model Matrix (Experimental, Preview, GA)](https://docs.google.com/document/d/1pwkifJHcYKGxf8J2Swd5DcWPZ0OhwPgwAubwbEz7-Mw/edit?usp=sharing)

@bmorelli25 if we shouldn't be using `beta`, we should (or I can) update the examples in our `applies_to` [syntax guidance](https://elastic.github.io/docs-builder/syntax/applies/).

  stack: beta 9.5+
products:
  - id: observability
  - id: elasticsearch
  - id: kibana
  - id: cloud-serverless
  - id: cloud-hosted
  - id: cloud-enterprise
  - id: cloud-kubernetes
  - id: elastic-stack
---

# How Significant Events works [sig-events-how-it-works]

This page describes the pipeline behavior for engineers and adopters who need to understand how Significant Events processes data, what runs where, and how to trace results across the system.

The pipeline has four sequential phases, each building on the outputs of the previous:

> - [Phase 1: KI extraction](#sig-events-hiw-ki): LLM and programmatic analysis of stream logs

Reviewer: First instance

```suggestion
- [Phase 1: Knowledge Indicators (KIs) extraction](#sig-events-hiw-ki): LLM and programmatic analysis of stream logs
```

- [Phase 2: Rule generation](#sig-events-hiw-rules): LLM generates detection rules using KIs as input
- [Phase 3: Rule execution](#sig-events-hiw-execution): Alerting framework runs promoted rules continuously
- [Phase 4: Discovery](#sig-events-hiw-discovery): Detection, discovery, and triage workflows convert alert signals into confirmed Significant Events

> ```mermaid

Reviewer: This visual is nice, but the distance between the first and second phases is comically long (like Littlefoot's-neck-in-The-Land-Before-Time-long). It's also not easy to scroll through the flowchart after expanding it. Last thiing - there's something a lil funny going on with one of the cells 🫣 I've not worked with mermaid diagrams before so idk how to fix it:

<img width="247" height="154" alt="Image" src="https://github.com/user-attachments/assets/511ac3b3-4d90-416a-9b2c-89b96778ec25" />

If you can shorten the distance between the first and second phase, that'd be optimal. If that's not possible, it'd be good to revisit this flowchart down the line or use a different tool to generate it.

> ```mermaid

Reviewer 2: This diagram is probably a bit complex and doesn't render well in the narrow page, since it has to verticalize a lot as show in the picture.

Also, it exposes a lot of information like the target indices that, IMO, are not necessary for the high-level diagram and you specify the storage detail in each sub-section later, which is good enough. Removing them for starting would potentially simplify this already.

<img width="521" height="1481" alt="Image" src="https://github.com/user-attachments/assets/f04f9f92-92d4-4058-96d1-bed07617899c" />

Author: I've removed this for now, but I'll recreate outside of mermaid. formatting was still off.

    subgraph KB["Kibana"]
        KI["KI extraction<br/>Task Manager + Workflow"]
        RG["Rule generation<br/>Workflow"]
        RE["Rule execution<br/>Alerting framework"]
        DW["Detection<br/>Workflows · cron 10m"]
        DA["Discovery agent<br/>Workflows + Agent Builder"]
        JA["Judge agent<br/>Workflows + Agent Builder"]
    end

    subgraph ES["Elasticsearch"]
        LOGS[("stream logs")]
        FEAT[(".significant_events-knowledge_indicators")]
        ALRT[(".rule-events")]
        DETS[(".significant_events-detections")]
        DISC[(".significant_events-discoveries")]
        EVTS[(".significant_events-events")]
    end

    LOGS -->|samples| KI
    KI <-.->|LLM| LLM
    KI --> FEAT
    FEAT --> RG
    RG <-.->|LLM| LLM
    RG -->|promote rules| RE
    RE -->|queries| LOGS
    RE --> ALRT
    ALRT --> DW
    DW --> DETS
    DETS --> DA
    FEAT -.->|context| DA
    DA <-.->|LLM| LLM
    DA --> DISC
    DISC --> JA
    JA <-.->|LLM| LLM
    JA --> EVTS

    classDef llm fill:#5c3fa3,color:#fff,stroke:none
    classDef det fill:#1d6b47,color:#fff,stroke:none
    class KI,RG,DA,JA llm
    class RE,DW det
```

## Phase 1: KI extraction [sig-events-hiw-ki]

> Knowledge Indicators (KIs) are stable, evidence-backed facts about the services and infrastructure present in a stream's log data. The LLM that generates detection operates entirely on KIs. This keeps rule generation deterministic and decouples detection logic from data format specifics.

Reviewer: ```suggestion
KIs are stable, evidence-backed facts about the services and infrastructure present in a stream's log data. The LLM that generates detection operates entirely on KIs. This keeps rule generation deterministic and decouples detection logic from data format specifics.
```

### How KIs are extracted

The pipeline runs up to five iterations, sampling up to 20 documents per round from entity-filtered, diverse, and random buckets. KIs found in one iteration are excluded from the next, steering each round toward undiscovered patterns. LLM analysis and deterministic code generators run in parallel; deterministic KIs always score 100 confidence. See [KI extraction](./knowledge-indicators.md#sig-events-ki-extraction) for the full extraction flow.

### KI types

> KIs cover five types: Entity, Infrastructure, Technology, Dependency, and Schema. Entity KIs have the highest extraction priority — services and databases are identified before individual pods or hosts. See [KI types](./knowledge-indicators.md#sig-events-ki-types) for the full data model.

Reviewer 2: thought (non-blocking): "highest extraction priority" overstates what's in the code.

The system prompt does mark Entity as the "Primary Focus" and tells the LLM not to create entities for individual pods/hosts, so the spirit is right. But there's no code-level priority ordering that runs entity extraction first; the round-robin over already-identified features is actually sorted alphabetically (`identify_inferred_features.ts`). If you want to keep this, I'd frame it as prompt guidance ("the model is instructed to prioritize services and databases over individual instances") rather than an extraction-priority guarantee.

### KI lifecycle

> KIs are written to `.significant_events-knowledge_indicators`, deduplicated on `(type, subtype, properties)`, and expire after 30 days if not re-observed. See [KI lifecycles](./knowledge-indicators.md#sig-events-ki-lifecycle) for lifecycle and continuous extraction details.

Reviewer: Some minor adjustments for clarity. As usual, fix if wrong.

```suggestion
KIs are written to `.significant_events-knowledge_indicators` and deduplicated on `type`, `subtype`, and `properties`. KIs expire after 30 days if they aren't re-observed. See [KI lifecycles](./knowledge-indicators.md#sig-events-ki-lifecycle) for lifecycle and continuous extraction details.
```

## Phase 2: Rule generation [sig-events-hiw-rules]

> Once KIs exist for a stream, the LLM generates detection rules using them as input. Rules are expressed as native {{esql}} expressions targeting the stream's index pattern:

Reviewer: Made some adjustments because some of the language (e.g., "rules are expressed as") reads as implementation details. Hopefully this is more direct and still accurate: 

```suggestion
Once a stream has KIs, the LLM generates a detection rule scoped to that stream. The generated rule is an {{esql}} query with conditions derived from the stream's KIs:
```

```
FROM {stream},{stream}.* METADATA _id | WHERE <esql_expression>
```

The LLM operates in a reasoning loop: it calls `get_stream_features` to retrieve KIs for the stream, analyzes what entities and technologies are present, then calls `add_queries` to submit a batch of rules. This means rules are scoped to the specific services and failure modes present in each stream.

## Phase 3: Rule execution [sig-events-hiw-execution]

> Each promoted rule runs as a {{kib}} alerting rule on a per-rule schedule. On each execution cycle, the rule:

Reviewer: ```suggestion
Each promoted KI runs as a {{kib}} alerting rule, on its own schedule. On each execution cycle, the rule:
```

1. Builds an {{esql}} query from the stored rule parameters.
2. Queries the stream with a lookback window of 2× the rule interval to ensure no events fall through the gap between runs.
3. Excludes document IDs seen in previous executions.

> 4. Writes matching rows to `.rule-events`.

Reviewer 2: This will be obsolete as soon as https://github.com/elastic/kibana/pull/280339 gets merged in a few days, I left what's changed in the PR description so you can use it to update this section please

> **Execution cap**: Each rule execution writes at most 1,000 alert documents. Matching events beyond this limit are dropped.

Reviewer: Potential revision that adds a bit more detail to improve clarity:

```suggestion
**Write limits**: Each time a rule executes, it can write up to 1000 documents. Once this limit is met, documents aren't written to `.rule-events` until the next time the rule executes. 
```

## Phase 4: Discovery [sig-events-hiw-discovery]

Discovery converts raw alert signals into confirmed Significant Events. It runs as three sequenced workflows (detection, discovery, and triage) coordinated by an Orchestrator workflow.

> - [Detection Workflow](#sig-events-hiw-detection) — `change_point` analysis per alerting rule, written to the detections index

Reviewer: ```suggestion
- [Detection Workflow](#sig-events-hiw-detection) — Change point aggregation per alerting rule, written to the detections index
```

> - [Discovery Workflow](#sig-events-hiw-discovery-agent) — discovery agent generates hypotheses, written to the discoveries index

Reviewer: ```suggestion
- [Discovery workflow](#sig-events-hiw-discovery-agent) — Discovery agent generates hypotheses, written to the discoveries index
```

> - [Triage Workflow](#sig-events-hiw-triage) — judge agent independently verifies and promotes, written to the events index

Reviewer: ```suggestion
- [Triage workflow](#sig-events-hiw-triage) — Judge agent independently verifies and promotes, written to the events index
```

### Detection [sig-events-hiw-detection]

> The Detection Workflow reads `.rule-events` and runs `change_point` analysis grouped per alerting rule. When a rule's alert pattern enters a genuinely anomalous state, the workflow appends a detection document to the detections index. Current state per rule is the latest document.

Reviewer: This took me some time to understand. I queried the nightshift program repo to answer my questions and came up with the following revision:

```suggestion
A Detection workflow reads `.rule-events`, buckets each rule's alerts by time, and runs the change point aggregation function against those counts to detect statistically significant shifts. When a shift is found, a document is appended to the detections index. The most recent detection document for a rule reflects its current state, for example, whether it's still producing an unusual rate of alerts or has returned to normal.
```

> Change point detection is per-rule: a stream can have many independent rules, and one rule recovering does not collapse the signal from others still anomalous on the same stream.

Reviewer: ```suggestion
Change point detection is per-rule. A stream can have many independent rules, and one rule recovering does not collapse the signal from others still anomalous on the same stream.
```

### Discovery agent [sig-events-hiw-discovery-agent]

> The discovery workflow reads unhandled detection documents and calls the discovery agent. The discovery agent is a hypothesis agent. It reads detection signals and produces structured discovery documents describing what is happening. Its scope is observation only, describing the anomaly, not prescribing remediation.

Reviewer: ```suggestion
A Discovery workflow reads unhandled detection documents and calls the Discovery agent. The Discovery agent is a hypothesis agent. It reads detection signals and produces structured discovery documents describing what is happening. Its scope is observation only, describing the anomaly, not prescribing remediation.
```

> ### Triage and the judge agent [sig-events-hiw-triage]

Reviewer: ```suggestion
### Triage and the Judge agent [sig-events-hiw-triage]
```

> The triage workflow reads unassessed discovery documents and calls the judge agent. The judge agent independently verifies the discovery agent's hypothesis and promotes, acknowledges, demotes, or resolves it. The judge agent runs in a separate invocation with no shared state. It receives the discovery agent's output as input but is explicitly instructed to treat it as a hypothesis to challenge, not a conclusion to ratify. Status transitions are written as new documents to the events index.

Reviewer: ```suggestion
The Triage workflow reads unassessed discovery documents and calls the Judge agent. The Judge agent independently verifies the Discovery agent's hypothesis and promotes, acknowledges, demotes, or resolves it. The Judge agent runs in a separate invocation with no shared state. It receives the Discovery agent's output as input but is explicitly instructed to treat it as a hypothesis to challenge, not a conclusion to ratify. Status transitions are written as new documents to the events index.
```

> The triage workflow reads unassessed discovery documents and calls the judge agent. The judge agent independently verifies the discovery agent's hypothesis and promotes, acknowledges, demotes, or resolves it. The judge agent runs in a separate invocation with no shared state. It receives the discovery agent's output as input but is explicitly instructed to treat it as a hypothesis to challenge, not a conclusion to ratify. Status transitions are written as new documents to the events index.

Reviewer 2: issue (blocking): same status-verb problem as `index.md`.

"promotes, acknowledges, demotes, or resolves it" — none of those four are real. The statuses are `open` / `closed` / `dismissed`. Suggest:

```suggestion
The triage workflow reads unassessed discovery documents and calls the judge agent. The judge agent independently verifies the discovery agent's hypothesis and sets the event status to open (page on-call), closed (confirmed settled), or dismissed (low severity and low confidence). The judge agent runs in a separate invocation with no shared state. It receives the discovery agent's output as input but is explicitly instructed to treat it as a hypothesis to challenge, not a conclusion to ratify. Status transitions are written as new documents to the events index.
```

Each promoted rule runs as a {{kib}} alerting rule on a per-rule schedule. On each execution cycle, the rule:

1. Builds an {{esql}} query from the stored rule parameters.
2. Queries the stream with a lookback window of 2× the rule interval to ensure no events fall through the gap between runs.
3. Excludes document IDs seen in previous executions.
4. Writes matching rows to `.rule-events`.

**Execution cap**: Each rule execution writes at most 1,000 alert documents. Matching events beyond this limit are dropped.

## Phase 4: Discovery [sig-events-hiw-discovery]

Discovery converts raw alert signals into confirmed Significant Events. It runs as three sequenced workflows (detection, discovery, and triage) coordinated by an Orchestrator workflow.

- [Detection Workflow](#sig-events-hiw-detection) — `change_point` analysis per alerting rule, written to the detections index
- [Discovery Workflow](#sig-events-hiw-discovery-agent) — discovery agent generates hypotheses, written to the discoveries index
- [Triage Workflow](#sig-events-hiw-triage) — judge agent independently verifies and promotes, written to the events index

### Detection [sig-events-hiw-detection]

> The Detection Workflow reads `.rule-events` and runs `change_point` analysis grouped per alerting rule. When a rule's alert pattern enters a genuinely anomalous state, the workflow appends a detection document to the detections index. The index is append-only — `detection`, `quiet`, and `handled` document kinds track anomaly state over time. Current state per rule is the latest document.

Reviewer 3: detection`, `quiet`, and `handled` document kinds have been removed. detection docs don't have kind

The triage workflow reads unassessed discovery documents and calls the judge agent. The judge agent independently verifies the discovery agent's hypothesis and promotes, acknowledges, demotes, or resolves it. The judge agent runs in a separate invocation with no shared state. It receives the discovery agent's output as input but is explicitly instructed to treat it as a hypothesis to challenge, not a conclusion to ratify. Status transitions are written as new documents to the events index.

The two-agent design is intended to prevent a single agent that both investigates and judges from confirming its own hypothesis.

## {{esql}} traceability [sig-events-hiw-traceability]

{{esql}} is the exclusive query language across the pipeline. Every rule is expressed as {{esql}}. Every artifact is stored as an {{es}} document. This means any point in the pipeline is queryable without leaving {{esql}}.

| What to query | {{esql}} target |
|---|---|
| Knowledge Indicators | `FROM .significant_events-knowledge_indicators` |
| Alert documents (rule execution results) | `FROM .rule-events` |
| Detection state per rule | `FROM .significant_events-detections` |

> | Discovery hypotheses | `FROM .significant_events-discoveries` |

Reviewer 3: We're planning to remove this. Should we mention it in the docs?

| Significant Events | `FROM .significant_events-events` |

> **Traceability**: The `kibana.alert.rule.uuid` in alert documents links back to the rule definition stored in the stream. The `kibana.alert.rule.tags` array always includes the stream name, so you can filter the alerts index by stream without a join. Across the discovery phase, `discovery_slug` is the durable cross-index key linking discoveries to events. To trace a Significant Event back to its origin, follow `discovery_slug` from the events index to the discoveries index, then use `detections[].detection_id` to reach the detection document, and `kibana.alert.rule.uuid` to reach the alerting rule.

Reviewer 3: This is outdated:
- instead of `discovery_slug` we should use `event_id`
- `detections[].detection_id` -> `signals[].metadata.detection_id` (this is used to connect to `.significant_events-detections`
- to reach the alert index: `signals[].metadata.rule_uuid` 

Should we mention the alert index here?

> **Traceability**: The `kibana.alert.rule.uuid` in alert documents links back to the rule definition stored in the stream. The `kibana.alert.rule.tags` array always includes the stream name, so you can filter the alerts index by stream without a join. Across the discovery phase, `discovery_slug` is the durable cross-index key linking discoveries to events. To trace a Significant Event back to its origin, follow `discovery_slug` from the events index to the discoveries index, then use `detections[].detection_id` to reach the detection document, and `kibana.alert.rule.uuid` to reach the alerting rule.

Reviewer 3: Should we add examples of queries to match `events` with `detections`?
